### PR TITLE
Aws waf region

### LINKED ISSUE #44767: AWS Waf-Regional support in waf modules
##### SUMMARY
waf modules do not currently support waf-regional (see boto: https://boto3.readthedocs.io/en/latest/reference/services/waf-regional.html).

Additionally, It's also a bit confusing that waf modules usually take a "region" parameter which is the connection region for aws.  It also might be worth clarifying the global description of the region parameter for AWS modules to note that this is for the boto connection, not for a region option for any given module.

##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
aws_waf_condition_module

##### ANSIBLE VERSION

```
ansible 2.6.2
  config file = /home/jackson/src/blend-ansible/ansible.cfg
  configured module search path = ['/home/jackson/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.5 (default, Apr  1 2018, 05:46:30) [GCC 7.3.0]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### STEPS TO REPRODUCE
N/A

##### EXPECTED RESULTS
Either an aws_wafregional_condition task type or an option like:
```
- aws_waf_condition:
    name: my_condition
    filter:
       - country: US
    type: geo
    alb_region: us-east-1 (default: 'global' or 'cloudfront')
```

### PR DESCRIPTION
**Pull request by mjmayer**
##### SUMMARY
Adds support for WAF Regional to the Ansible WAF Modules

Fixes #44767 #44896
##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
aws_waf_condition
aws_waf_facts
aws_waf_rule
aws_waf_web_acl
module_utils/aws/waf
module_utils/aws/waiters

##### ADDITIONAL INFORMATION
There are two WAF modules in botocore. WAF and WAFRegional. They function largely in the same way. But there are differences. These changes add support for WAFRegional without changing the existing behavior of the module.

Previously WAF worked like:
```
- name: create IP address condition
  aws_waf_condition:
    name: "{{ resource_prefix }}_ip_condition"
    filters:
      - ip_address: "10.0.0.0/8"
      - ip_address: "192.168.0.0/24"
    type: ip
```

It now supports WAF Regional and takes the waf_regional bool parameter
```
- name: create IP address condition
  aws_waf_condition:
    name: "{{ resource_prefix }}_ip_condition"
    filters:
      - ip_address: "10.0.0.0/8"
      - ip_address: "192.168.0.0/24"
    type: ip
    waf_regional: true
    region: '{{ aws_region }}'
```

### CODE REVIEW

**Comment by davehewy** [comment]
@mjmayer well done for raising the PR! Looks v.v.similar to my working branch :+1:

**Comment by davehewy** [comment]
@mjmayer want a hand with tidying this pr up at all?

**Comment by mjmayer** [comment]
@DaveHewy Sure, if you think it can be improved on. I'll take the help.

@Sivel Could you look over this PR and give feedback if any changes are needed?

**Comment by mjmayer** [comment]
@sivel are you able to review this PR?

File: lib/ansible/modules/cloud/amazon/aws_waf_condition.py [hunk 1]
```diff
@@ -57,6 +57,12 @@
         - Whether to remove existing filters from a condition if not passed in I(filters).
         default: False
         type: bool
+    waf_regional:
+        description: Wether to use waf_regional module. Defaults to false.
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_condition.py]: typo: whether

**Comment by willthames** [comment]
Other than the 'Wether' typo, this looks good to me.

**Comment by davehewy** [comment]
@mjmayer @willthames @sivel  what is now required to get this over the line?

**Comment by DaveHewy** [comment]
> Other than the 'Wether' typo, this looks good to me.

@willthames - can we get merging then?

**Comment by mjmayer** [comment]
@willthames Any chance we can get some love on adding support for the regional wafs and get it merged?

**Comment by mjmayer** [comment]
@sivel This PR is good to go. What is preventing it from getting merged?

File: lib/ansible/modules/cloud/amazon/aws_waf_condition.py [hunk 2]
```diff
@@ -57,6 +57,12 @@
         - Whether to remove existing filters from a condition if not passed in I(filters).
         default: False
         type: bool
+    waf_regional:
+        description: Whether to use waf_regional module. Defaults to false.
+        default: false
+        required: no
+        type: bool
+        version_added: 2.8
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_condition.py]: Needs to be 2.9 now

File: lib/ansible/modules/cloud/amazon/aws_waf_facts.py [hunk 3]
```diff
@@ -17,6 +17,12 @@
   name:
     description:
       - The name of a Web Application Firewall
+  waf_regional:
+      description: Wether to use waf_regional module. Defaults to true
+      default: false
+      required: no
+      type: bool
+      version_added: "2.8"
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_facts.py]: 2.9 now

File: lib/ansible/modules/cloud/amazon/aws_waf_rule.py [hunk 4]
```diff
@@ -145,8 +151,21 @@ def get_rule(client, module, rule_id):
...
+    if client.__class__.__name__ == 'WAF':
+        try:
+            return list_rules_with_backoff(client)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg='Could not list WAF rules')
+    elif client.__class__.__name__ == 'WAFRegional':
+        try:
+            return list_regional_rules_with_backoff(client)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg='Could not list WAF rules')
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_rule.py]: Is it worth saying regional rules here?

File: lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py [hunk 5]
```diff
@@ -49,12 +49,17 @@
...
         description:
         - Whether to remove rules that aren't passed with C(rules).
         default: False
         type: bool
+    waf_regional:
+        description: Wether to use waf_regional module. Defaults to false.
+        default: false
+        required: no
+        type: bool
+        version_added: "2.8"
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py]: 2.9

File: lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py [hunk 6]
```diff
@@ -49,12 +49,17 @@
...
         - Priorities must be unique, but not necessarily consecutive. Lower numbered priorities are evalauted first.
         - The I(type) key can be passed as C(rate_based), it defaults to C(regular)
-
     purge_rules:
         description:
         - Whether to remove rules that aren't passed with C(rules).
         default: False
         type: bool
+    waf_regional:
+        description: Wether to use waf_regional module. Defaults to false.
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py]: typo: whether

File: lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py [hunk 7]
```diff
@@ -152,11 +158,18 @@ def get_web_acl_by_name(client, module, name):
...
+            rules = list_rules_with_backoff(client)
+            return dict((rule['Name'], rule) for rule in rules)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg='Could not list rules')
+    elif client.__class__.__name__ == 'WAFRegional':
+        try:
+            rules = list_regional_rules_with_backoff(client)
+            return dict((rule['Name'], rule) for rule in rules)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg='Could not list rules')
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py]: regional rules?

File: lib/ansible/modules/cloud/amazon/aws_waf_rule.py [hunk 8]
```diff
@@ -47,8 +47,14 @@
     purge_conditions:
         description:
           - Whether or not to remove conditions that are not passed when updating `conditions`.
-        default: False
+        default: false
         type: bool
+    waf_regional:
+        description: Wether to use waf_regional module. Defaults to false
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_rule.py]: typo: Whether

File: lib/ansible/modules/cloud/amazon/aws_waf_rule.py [hunk 9]
```diff
@@ -47,8 +47,14 @@
...
+        default: false
+        required: no
+        type: bool
+        version_added: "2.8"
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_waf_rule.py]: 2.9 now

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
Apologies for not revisiting this sooner - I don't always see mentions. Feel free to nudge me on IRC for re-reviews (nudging here does work, as seen now, but it less likely)

File: lib/ansible/modules/cloud/amazon/aws_waf_condition.py (hunk 2 above)
Comment (mjmayer) [reply lib/ansible/modules/cloud/amazon/aws_waf_condition.py]: 48288e1

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
Merged, thanks @mjmayer! Apologies again for the various delays

**Pull request merged**